services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedDeviceBackend.java

```diff
@@ -161,23 +159,23 @@ public MongoDbBasedDeviceBackend(
                 .compose(result -> {
                     if (result.getStatus() == HttpURLConnection.HTTP_NOT_FOUND
                             && isAutoProvisioningEnabled(type, clientContext)) {
-                        try {
-                            final X509Certificate cert = DeviceRegistryUtils
-                                    .getCertificateFromClientContext(tenantId, authId, clientContext, span);
-                            return provisionDevice(tenantId, cert, span)
-                                    .compose(r -> {
-                                        if (r.isError()) {
-                                            TracingHelper.logError(span, r.getPayload());
-                                            return Future.succeededFuture(
-                                                    createErrorCredentialsResult(r.getStatus(), r.getPayload()));
-                                        } else {
-                                            return getNewCredentials(tenantId, authId, span);
-                                        }
-                                    });
-                        } catch (final ClientErrorException e) {
-                            return Future
-                                    .succeededFuture(createErrorCredentialsResult(e.getErrorCode(), e.getMessage()));
-                        }
+                        return DeviceRegistryUtils
+                                .getCertificateFromClientContext(tenantId, authId, clientContext, span)
+                                .compose(optionalCert -> optionalCert
+                                        .map(ok -> provisionDevice(tenantId, optionalCert.get(), span)
+                                                .compose(r -> {
+                                                    if (r.isError()) {
+                                                        TracingHelper.logError(span, r.getPayload());
+                                                        return Future.succeededFuture(
+                                                                createErrorCredentialsResult(r.getStatus(),
+                                                                        r.getPayload()));
+                                                    } else {
+                                                        return getNewCredentials(tenantId, authId, span);
+                                                    }
+                                                }))
+                                        .orElse(Future.succeededFuture(
+                                                createErrorCredentialsResult(HttpURLConnection.HTTP_BAD_REQUEST,
+                                                        "Not able to get the certificate from the client context"))));
```

Comment:
why is it a bad request if the client context does not contain a certificate? Maybe the adapter doesn't want to support auto provisioning?

Reply:
I see the point. If the client context does not contain a certificate, instead of `BAD_REQUEST`, I will pass the original result which is `HTTP_NOT_FOUND` as below.
`orElse(Future.succeededFuture(result)));`